services/src/datasets/storage.rs

```diff
@@ -0,0 +1,188 @@
+use crate::datasets::listing::DataSetProvider;
+use crate::error::Result;
+use crate::projects::project::LayerInfo;
+use crate::users::user::UserId;
+use crate::util::user_input::{UserInput, Validated};
+use async_trait::async_trait;
+use geoengine_datatypes::dataset::{DataSetId, DataSetProviderId, InternalDataSetId};
+use geoengine_datatypes::util::Identifier;
+use geoengine_operators::mock::MockDataSetDataSourceLoadingInfo;
+use geoengine_operators::source::OgrSourceDataset;
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+use std::path::PathBuf;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct DataSet {
+    pub id: DataSetId,
+    pub name: String,
+    pub description: String,
+    pub data_type: LayerInfo, // TODO: custom type instead of reusing existing one?
+    pub source_operator: String, // TODO: enum?
+}
+
+impl From<AddDataSet> for DataSet {
+    fn from(value: AddDataSet) -> Self {
+        Self {
+            id: DataSetId::Internal(InternalDataSetId::new()),
+            name: value.name,
+            description: value.description,
+            data_type: value.data_type,
+            source_operator: value.source_operator,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct AddDataSet {
+    pub name: String,
+    pub description: String,
+    pub data_type: LayerInfo,    // TODO: custom type?
+    pub source_operator: String, // TODO: enum?
+}
+
+impl UserInput for AddDataSet {
+    fn validate(&self) -> Result<()> {
+        // TODO
+        Ok(())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct ImportDataSet {
+    pub name: String,
+    pub description: String,
+    pub data_type: LayerInfo,
+    pub source_operator: String,
+}
+
+impl From<ImportDataSet> for DataSet {
+    fn from(value: ImportDataSet) -> Self {
+        DataSet {
+            id: DataSetId::Internal(InternalDataSetId::new()),
+            name: value.name,
+            description: value.description,
+            data_type: value.data_type,
+            source_operator: value.source_operator,
+        }
+    }
+}
+
+impl UserInput for ImportDataSet {
+    fn validate(&self) -> Result<()> {
+        // TODO
+        Ok(())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct DataSetProviderListing {
+    pub id: DataSetProviderId,
+    pub name: String,
+    pub description: String,
+    // more meta data (number of data sets, ...)
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub enum AddDataSetProvider {
+    AddMockDataSetProvider(AddMockDataSetProvider),
+    // TODO: geo catalog, wcs, ...
+}
```

Comment:
Also here, maybe consider typetag?

Reply:
it might make sense here but I would add it as a TODO because I am not sure yet how to implement the providers at all. I would address it in the PR that actually implements the backend.